File: test_upload_error.js
#!/usr/bin/env node

'use strict';

const request = require('request');
const nconf = require('nconf');
const path = require('path');

// Load NodeBB configuration
nconf.file({
    file: path.join(__dirname, 'config.json')
});

nconf.defaults({
    url: 'http://127.0.0.1:4567'
});

const helpers = require('./test/helpers');

async function testUploadError() {
    try {
        // Start the server
        console.log('Starting NodeBB server...');
        const child = require('child_process').spawn('node', ['loader.js', '--no-silent'], {
            cwd: __dirname,
            stdio: 'inherit'
        });

        // Wait for server to start
        await new Promise(resolve => setTimeout(resolve, 5000));

        // Create admin user
        const adminUid = await new Promise((resolve, reject) => {
            require('./src/user').create({ username: 'admin', password: 'adminpass' }, (err, uid) => {
                if (err) reject(err);
                else resolve(uid);
            });
        });

        // Make user admin
        await new Promise((resolve, reject) => {
            require('./src/groups').join('administrators', adminUid, (err) => {
                if (err) reject(err);
                else resolve();
            });
        });

        // Login as admin
        const { jar, csrf_token } = await helpers.loginUser('admin', 'adminpass');

        console.log('\nTesting upload with invalid file type...');
        
        // Try to upload a text file as favicon (should fail - only .ico allowed)
        const testFile = path.join(__dirname, 'test/files/test.txt');
        
        // Create test file if it doesn't exist
        require('fs').writeFileSync(testFile, 'test content');

        // Attempt upload
        await new Promise((resolve) => {
            const formData = {
                'files[]': require('fs').createReadStream(testFile),
                params: JSON.stringify({})
            };

            request.post({
                url: `${nconf.get('url')}/api/admin/uploadfavicon`,
                formData: formData,
                jar: jar,
                headers: {
                    'x-csrf-token': csrf_token
                },
                json: true
            }, (err, res, body) => {
                console.log('\nResponse Status Code:', res.statusCode);
                console.log('Response Body:', JSON.stringify(body, null, 2));
                
                if (res.statusCode === 200) {
                    console.log('\n❌ ERROR: Server returned 200 OK for invalid upload!');
                    console.log('Expected: 4xx or 5xx error status code');
                } else {
                    console.log('\n✓ Correct: Server returned error status code');
                }
                
                resolve();
            });
        });

        // Kill the server
        child.kill();
        process.exit(0);
    } catch (err) {
        console.error('Test failed:', err);
        process.exit(1);
    }
}

// Run the test
testUploadError();

File: test_upload_error.py
#!/usr/bin/env python3

import requests
import json
import tempfile
import os

# This script tests the admin upload functionality to reproduce the error
# described in the PR - when uploading an invalid file type, the server
# responds with HTTP 200 instead of an error status code

def test_admin_upload_error():
    """
    Test uploading an invalid file type to admin upload endpoints
    to verify that HTTP 200 is returned even when errors occur
    """
    
    # Create a test file with invalid content
    with tempfile.NamedTemporaryFile(mode='w', suffix='.txt', delete=False) as f:
        f.write('This is a text file, not an image')
        test_file_path = f.name
    
    print(f"Created test file: {test_file_path}")
    
    # Test various admin upload endpoints
    base_url = "http://localhost:4567"
    endpoints = [
        '/api/admin/uploadCategoryPicture',
        '/api/admin/uploadFavicon', 
        '/api/admin/uploadTouchIcon',
        '/api/admin/uploadMaskableIcon',
        '/api/admin/uploadLogo'
    ]
    
    try:
        for endpoint in endpoints:
            print(f"\nTesting endpoint: {endpoint}")
            
            # Prepare the file upload
            with open(test_file_path, 'rb') as file_content:
                files = {'files': (os.path.basename(test_file_path), file_content, 'text/plain')}
                params = {'cid': 1} if 'Category' in endpoint else {}
                data = {'params': json.dumps(params)} if params else {}
                
                try:
                    response = requests.post(
                        base_url + endpoint,
                        files=files,
                        data=data,
                        timeout=5
                    )
                    
                    print(f"Status Code: {response.status_code}")
                    print(f"Response Headers: {dict(response.headers)}")
                    
                    try:
                        response_json = response.json()
                        print(f"Response Body: {json.dumps(response_json, indent=2)}")
                        
                        # Check for the specific issue: HTTP 200 with error in body
                        if response.status_code == 200 and 'error' in response_json:
                            print("❌ BUG CONFIRMED: HTTP 200 returned with error in body!")
                            print(f"Error message: {response_json['error']}")
                        elif response.status_code >= 400:
                            print("✅ Expected behavior: HTTP error status returned")
                        else:
                            print("? Unexpected response format")
                            
                    except json.JSONDecodeError:
                        print(f"Response Text: {response.text}")
                        
                except requests.exceptions.ConnectionError:
                    print(f"❓ Cannot connect to {base_url} - server may not be running")
                except requests.exceptions.Timeout:
                    print("❓ Request timed out")
                except Exception as e:
                    print(f"❓ Error making request: {e}")
    
    finally:
        # Clean up test file
        if os.path.exists(test_file_path):
            os.unlink(test_file_path)
            print(f"\nCleaned up test file: {test_file_path}")

if __name__ == "__main__":
    test_admin_upload_error()